### PR TITLE
Add Json nesting depth calculation during parsing 

### PR DESCRIPTION
**Pull request by sandwwraith**
+ Set Deprecated(HIDDEN) to setters in JsonConfiguration
+ Increase stack size for mingwx64 test binaries

### CODE REVIEW

**Comment by sandwwraith** [comment]
#3029 can be merged after this

File: formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2017-2026 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.*
```
Comment (fzhinkin) [inline formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt]:
```suggestion
```

File: formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import kotlinx.serialization.json.JsonTestBase
+import kotlinx.serialization.json.JsonTestingMode
+import kotlinx.serialization.test.checkDecodingException
+import kotlin.test.Test
+
+class NestingJsonTest: JsonTestBase() {
```
Comment (fzhinkin) [inline formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt:16]: Please also add tests verifying behavior one level before hitting the limit and just hitting it (I mean, 499-depth object should be parsed, 500-depth object should be rejected).

And it would be nice to check that the limit could be actually adjusted by the corresponding property.

**Review by fzhinkin — CHANGES_REQUESTED** [review]

File: formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt (hunk 2 above)
Comment (sandwwraith) [reply formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt:16]: > And it would be nice to check that the limit could be actually adjusted by the corresponding property.

This is verified [here](https://github.com/Kotlin/kotlinx.serialization/pull/3246/changes#diff-f0fc442f1356a7c486c42cc3361e0ddc904a43f0fddf42246cf2db3aa485a7e3R50)

**Comment by sandwwraith** [comment]
Apparently, #3235 added some stack variables, so now it's SoE with a 500 limit... gonna lower it more.

File: formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonTreeReader.kt [hunk 3]
```diff
@@ -110,9 +119,27 @@ internal class JsonTreeReader(
...
+            TC_BEGIN_LIST -> withDepthControl { readArray() }
             else -> lexer.fail("Can't begin reading element, unexpected token")
         }
     }.invoke(Unit)
+
+    private inline fun withDepthControl(read: () -> JsonElement): JsonElement {
+        if (++stackDepth > maxNestingDepth) {
+            jsonTooNested(maxNestingDepth)
+        }
+        val result = read()
```
Comment (shanshin) [inline formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonTreeReader.kt:135]: Am I right that if an exception is thrown in `read()`, we’ll go back to the very beginning of the serialization and won’t access `stackDepth` again?

Comment (sandwwraith) [reply formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonTreeReader.kt:135]: Yes, any exception from lexer/parser should completely stop deserialization and unwind all stack

File: formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt (hunk 2 above)
Comment (shanshin) [reply formats/json-tests/commonTest/src/kotlinx/serialization/NestingJsonTest.kt:16]: I also think we should check the boundary condition, like  `maxNestingDepth = 2` and JSON with depth 2 should be parsed successfully

**Review by shanshin — APPROVED** [review]

**Review by fzhinkin — APPROVED** [review]
